functions.lua

```diff
@@ -362,6 +433,131 @@ function M.get_crossref(str)
   return ret
 end
 
+-- function for print table values-keys recursively for debugging
+function M.printTable(t,l)
+ local nextlevel=l+1
+ for k,v in pairs(t) do
+  if type(v)=='table' then
+   M.printTable(v,nextlevel)
+  else
+   print (l..'. '..k..': '..v)
+  end
+ end
+end
+
+-- function for trying to convert (arXiv API returned) titles properly case-preserving string by BibTeX
+-- + converting greek letters and leq/geq chars in titles by their TeX commands
+function M.ucaseTitle(t)
+ local to=t:gsub('(%$[^%$]+%$)','{%1}')
+ if (t:find('%s%l%l%l%l%l')) then -- there exists at least one at least 5 character long lowercase word
+   to=to:gsub('(%u+)','{%1}')
+ end
+ local Greeks = {'alpha','beta','Gamma','gamma','Delta','delta','epsilon','varepsilon',
+  'zeta','eta','Theta','theta','vartheta','iota','kappa','Lambda','lambda',
+  'mu','nu','xi','Xi','Pi','pi','rho','varrho','Sigma','sigma','varsigma',
+  'tau','Upsilon','upsilon','Phi','phi','varphi','chi','Psi','psi','Omega','omega',
+  'leq','geq'}
+ local Grchars = {'α','β','Γ','γ','Δ','δ','ϵ','ε',
+  'ζ','η','Θ','θ','ϑ','ɩ','κ','Λ','λ',
+  'μ','ν','ξ','Ξ','Π','π','ρ','ϱ','Σ','σ','ς',
+  'τ','Υ','υ','Φ','φ','ϕ','χ','Ψ','ψ','Ω','ω',
+  '≤','≥'}
+ for i = 1, #Greeks do
+  to=to:gsub(Grchars[i]..'%a','\\'..Greeks[i]..' ')
+  to=to:gsub(Grchars[i],'\\'..Greeks[i])
+ end
+ return to
+end
+
+--- query arXiv API for ID
+-- Input: string str
+-- Output: Atom XML -> table -> BibTeX data list
+function M.get_arxiv(str,str_naked)
+  local ret = ''
+  local arxiv=''
+  local axret=''
+  local axentry
+  -- trying to get arXiv ID from preprint \bibitem entries...
+  if str:find('[aA][rR][Xx][iI][vV][:%.%(]?%s*[Pp]?r?e%-?print[:%.]?%s*[%a%d%/%.][%a%d%/%.][%a%d%/%.][%a%d%/%.][%a%d%/%.][%a%d%/%.][%a%d%/%.]+%d') then
```

Comment:
i think this code would be easer to understand like this:

```suggestion
  local lower = str:lower()
  if lower:find('arxiv[:%.%(]?%s*p?r?e%-?print[:%.]?%s*[%a%d%/%.][%a%d%/%.][%a%d%/%.][%a%d%/%.][%a%d%/%.][%a%d%/%.][%a%d%/%.]+%d') then
```

Also, I think the regex is broken. At the moment it matches:

- `…-pre-print…`
- `…-re-print…`
- `…-pe-print…`
- `…-e-print…`
- `…-eprint…`

… but the `e` always stays. it also looks different to the regex in line 482, is that intentional?

Reply:
Thanks, I implemented your method (to convert string to lowercase for finding matches), not just here but also in bibcheck.lua, too.
I also updated the gsub code below for matching arXiv IDs with (pr)eprint inbetween to replicate the find above, thank you.

Reply:
(see the changes of the latest commit, also the bibcheck.lua, of the PR)